### PR TITLE
Add LTP support

### PR DESCRIPTION
**Pull request by ttnie**
## Summary
Add LTP support

## Testing
test on local environment

### CODE REVIEW

File: testing/ltp/Make.defs
Comment (xiaoxiang781216) [inline testing/ltp/Make.defs:2]: replace the copyright

File: testing/ltp/Kconfig
Comment (xiaoxiang781216) [inline testing/ltp/Kconfig:6]: change the name of option

**Review by xiaoxiang781216 — CHANGES_REQUESTED** [review]

File: testing/ltp/Kconfig
Comment (ttnie) [reply testing/ltp/Kconfig]: @xiaoxiang781216  update done.

File: testing/ltp/Make.defs
Comment (ttnie) [reply testing/ltp/Make.defs:2]: @xiaoxiang781216 update done

Comment (acassis) [reply testing/ltp/Make.defs:2]: Strange, it still displaying BSD License instead of Apache, is it some issue on github?

**Review by acassis — CHANGES_REQUESTED** [review]

File: testing/ltp/Make.defs
Comment (ttnie) [reply testing/ltp/Make.defs]: @acassis Thanks for your reminder, update done.

**Comment by acassis** [comment]
@ttnie is it ready to merge?

**Comment by btashton** [comment]
I would want to know more about the 20k line patch file before merging this in. It's GPL code in the patch.

**Comment by acassis** [comment]
Since it will be used only for testing and will not be included in the final product's firmware, maybe it is not a big concern. What you think?

**Comment by xiaoxiang781216** [comment]
> It's 20k lines of GPL we cannot have that in the Apache release bundle. Also it's not clear to me how we would maintain it.

The change isn't specific to NuttX, @ttnie and @PeterBee97 is preparing to upstream the patch to mainline.

**Comment by xiaoxiang781216** [comment]
> @ttnie is it ready to merge?

There are few compiling warnings need to resolve before merging. We ares in China Nation Holiday, the progress is a little bit slow.

**Comment by btashton** [comment]
> > It's 20k lines of GPL we cannot have that in the Apache release bundle. Also it's not clear to me how we would maintain it.
> 
> @btashton all changes aren't specific to NuttX, @ttnie and @PeterBee97 is preparing to upstream the patch to mainline.

Awesome that would be the best. If we hit a roadblock maybe we can consider handling it like we do with the other restricted tools. This is certainly something we want to be able to support

**Comment by acassis** [comment]
> > @ttnie is it ready to merge?
> 
> @acassis there are few compiling warnings need to resolve before merging. We ares in China Nation Holiday, the progress is a little bit slow.

Hi @xiaoxiang781216 I'm sorry, I forgot about it. When people will return? By Thursday?

**Comment by xiaoxiang781216** [comment]
Friday, but people may take one ro two days vacation.

**Comment by justinmclean** [comment]
Just trying to understand the GPL comment above. The GPL license is not compatible with the Apache one and we can't have GPL code in our repo. ASF projects also can't have a dependancy on GPL code either. Is this the case with this PR?

**Comment by btashton** [comment]
The this PR would enable users to optionally include the LTP as an application. To do so requires a very large patch applied (included in the PR) to what the user would choose to download.

That is why I had raised the issue about accepting this PR. The contents of the patch are being upstreamed to the LTP project. Once that is done we should be able to accept this as the GPL code will only be included if the user chooses to download it in addition to our Apache release.

**Comment by patacongo** [comment]
@justinmclean Can you please clarify this.  Particularly the phrase "ASF projects also can't have a dependancy on GPL code either."
> 
> Just trying to understand the GPL comment above. The GPL license is not compatible with the Apache one and we can't have GPL code in our repo. ASF projects also can't have a dependancy on GPL code either. Is this the case with this PR?

This could would download and build the GPL Linux Test Program at build time.  So, even if they work around the issues of the GPL patches, the binaries built with option would include GPL code.  Is this a "dependency on GPL".  It seems so to me.

There are several other places where this download trick is used to download third part code into NuttX binaries at build time.  This is done to avoid ASF third party rules, to avoid conformance with the NuttX coding standard, and to deal with third party code that is undergoing changes.  It would be good to understand if the ASF has any restrictions or thoughts on this practice -- with both compatible and incompatible licenses.

**Comment by btashton** [comment]
There are lot of other Apache projects that support being built with non allowed code such as some intel and Nvidia libraries. It is my understanding that the Apache project just still needs to be functional without it which certainly is the case here.

**Comment by patacongo** [comment]
> 
> 
> There are lot of other Apache projects that support being built with non allowed code such as some intel and Nvidia libraries. It is my understanding that the Apache project just still needs to be functional without it which certainly is the case here.

That seems like a reasonable interpretation to me.  But it would still be good to have an official ASF position.

**Comment by justinmclean** [comment]
Proprietary licenses like some Intel and Nvidia ones are different are in general not compatible with the Apache license as they add restrictions above and beyond what the Apache license allows. If a license add any condition (not just copyleft) then it cannot be included in a release or depended on.If you want an example of where a project can get into trouble look at the issues with MXNet project and it use of Nvidia licensed code. You can find this in the last several Incubator reports https://whimsy.apache.org/board/minutes/Incubator.html. The likely outcome of this is that either Nvidia will change it license or MXNet will no longer be an Apache project.
Category  X license dependancies are allowed if the feature is truely optional and the user need to install that themselves. See https://www.apache.org/legal/resolved.html#optional

**Comment by justinmclean** [comment]
Now in general using non compatible licenses for testing is fine as no incapable nothing ends up in the final source or binary release. e.g JUnit But it seem like this issue may be more similar with one of the MXNet issue it that GPL licensed code ends up in the binary. Is this correct? If so there is a dependency on GPL and would not be allowed unless it's an optional feature that is not built by default and the user is made aware of the licensing implication if they do enable it.

**Comment by justinmclean** [comment]
Here's the main information on GPL (and other Category X) licenses. https://www.apache.org/legal/resolved.html#prohibited

**Comment by patacongo** [comment]
> Category X license dependancies are allowed if the feature is truely optional and the user need to install that themselves. See https://www.apache.org/legal/resolved.html#optional

Certainly, this code is _optional_ by that definition.  In this case, the NuttX project code would provide the logic to automatically download the code without requiring that the user perform the download.  So the use of GPL is completely transparent.

**Comment by btashton** [comment]
Thanks @justinmclean yeah I had been watching the MXNet conversation.

Here I think we meet the Optional definition in that link. In this case a user would enable the LTP as an application specifically to test NuttX compliance with the POSIX interfaces, it would never be expected to be included in an end product using NuttX and no LTP GPL code would be included in the release.

We include some helpers in the build scripts to download the source but only if a user opts in to that feature.

The only users of this feature are likely NuttX contributors and maintainers.

**Comment by patacongo** [comment]
> 
> 
> Now in general using non compatible licenses for testing is fine as no incapable nothing ends up in the final source or binary release. e.g JUnit But it seem like this issue may be more similar with one of the MXNet issue it that GPL licensed code ends up in the binary. Is this correct? If so there is a dependency on GPL and would not be allowed unless it's an optional feature that is not built by default and the user is made aware of the licensing implication if they do enable it.

We don't release binaries and the GPL code would not be present in the source releases either. In the build process, the GPL source would be automatically downloaded on demand and built into the binaries by the user.  This code would not be downloaded or built in the _normal_ OS build; it would only be included the if the user were making test-related build and if the user explicitly selected inclusion of the GPL code from a configuration menu.

**Comment by justinmclean** [comment]
That all sounds fine to me, thanks for the explanation.

**Comment by ttnie** [comment]
> @ttnie is it ready to merge?

@acassis back from holiday.

**Comment by btashton** [comment]
@ttnie why do we need to make all of the functions and variables static here?  Seems like we should be able to work around this in the build without having to patch that in?

**Comment by ttnie** [comment]
> @ttnie why do we need to make all of the functions and variables static here? Seems like we should be able to work around this in the build without having to patch that in?

please ignore this PR, we'll apply LTP modifications to GPL not nuttx.

**Comment by btashton** [comment]
Ok, just wanted to see if there was a way to work around it if we face resistance upstream :)

**Comment by ttnie** [comment]
> @ttnie why do we need to make all of the functions and variables static here? Seems like we should be able to work around this in the build without having to patch that in?



> Ok, just wanted to see if there was a way to work around it if we face resistance upstream :)

I make sense, the modifications of LTP is plan to added to GPL by our team member. And no need patch at all.

**Comment by xiaoxiang781216** [comment]
> @ttnie why do we need to make all of the functions and variables static here? Seems like we should be able to work around this in the build without having to patch that in?

In Linux, each test case build into a separated ELF binary, but NuttX build all soruces into one big image(execpt we enable ELF binfmt). So the linker complain the duplicated symbols on NuttX. But anyway, it's always a good egnineering practice to limit the symbol scope as small as possible even it isn't so critical on Linux.

**Comment by btashton** [comment]
> In Linux, each test case build into a separated ELF binary, but NuttX build all soruces into one big image(execpt we enable ELF binfmt). So the linker complain the duplicated symbols on NuttX. But anyway, it's always a good egnineering practice to limit the symbol scope as small as possible even it isn't so critical on Linux.

Yeah I was wondering if we could just create standalone ELF for each, but I agree lets see if we can just improve the LTP and use that as a fallback otherwise.

File: testing/ltp/Makefile
Comment (xiaoxiang781216) [inline testing/ltp/Makefile:112]: need add distclean to keep git status clear.

Comment (ttnie) [reply testing/ltp/Makefile:116]: > need add distclean to keep git status clear.

add clean part to makefile, pls review

File: testing/ltp/Makefile
Comment (xiaoxiang781216) [inline testing/ltp/Makefile:113]: remove since .built isn't generated anymore

File: testing/ltp/Makefile
Comment (xiaoxiang781216) [inline testing/ltp/Makefile:112]: change to distclean

File: testing/ltp/Makefile
Comment (xiaoxiang781216) [inline testing/ltp/Makefile:114]: remove,

Comment (ttnie) [reply testing/ltp/Makefile]: > remove,

update to distclean and remove clean

**Comment by xiaoxiang781216** [comment]
LGTM, @btashton could you review again?

**Comment by jerpelea** [comment]
LGTM

**Pull request merged**
